network/src/protocols/discovery/addr.rs

```diff
@@ -52,50 +46,36 @@ pub trait AddressManager {
 
 // bitcoin: bloom.h, bloom.cpp => CRollingBloomFilter
 pub struct AddrKnown {
-    max_known: usize,
-    addrs: HashSet<RawAddr>,
-    addr_times: HashMap<RawAddr, Instant>,
-    time_addrs: BTreeMap<Instant, RawAddr>,
+    filters: StableBloomFilter<DefaultBuildHashKernels<RandomState>>,
 }
 
 impl AddrKnown {
     pub(crate) fn new(max_known: usize) -> AddrKnown {
         AddrKnown {
-            max_known,
-            addrs: HashSet::default(),
-            addr_times: HashMap::default(),
-            time_addrs: BTreeMap::default(),
+            filters: StableBloomFilter::new(
+                max_known,
+                3,
+                0.03,
+                DefaultBuildHashKernels::new(rand::random(), RandomState::default()),
+            ),
         }
     }
 
-    pub(crate) fn insert(&mut self, key: RawAddr) {
-        let now = Instant::now();
-        self.addrs.insert(key);
-        self.time_addrs.insert(now, key);
-        self.addr_times.insert(key, now);
-
-        if self.addrs.len() > self.max_known {
-            let first_time = {
-                let (first_time, first_key) = self.time_addrs.iter().next().unwrap();
-                self.addrs.remove(&first_key);
-                self.addr_times.remove(&first_key);
-                *first_time
-            };
-            self.time_addrs.remove(&first_time);
-        }
+    pub(crate) fn insert<T: ::std::hash::Hash>(&mut self, key: T) {
+        self.filters.insert(&key)
     }
 
-    pub(crate) fn contains(&self, addr: &RawAddr) -> bool {
-        self.addrs.contains(addr)
+    pub(crate) fn extend<'a, T: 'a + ::std::hash::Hash>(
+        &mut self,
+        keys: impl Iterator<Item = &'a T>,
+    ) {
+        for key in keys {
+            self.filters.insert(key)
```

Comment:
This is unrealizable. Bloomfilter's extend needs to traverse the hash once. Even if it provides a method called extend, it also traverses and inserts. 

And, as a table filter, when inserting, it will randomly delete an occupation, which means that it can ensure the existence of recent inserts as much as possible, and randomly delete more distant inserts